pkg/virt-launcher/virtwrap/network/network.go

```diff
@@ -63,42 +63,69 @@ type podNIC interface {
-	networks := map[string]*v1.Network{}
-	multusIndices := map[string]int{}
-	for _, network := range vmi.Spec.Networks {
-		networks[network.Name] = network.DeepCopy()
-		if networks[network.Name].Multus != nil && !networks[network.Name].Multus.Default {
-			// multus pod interfaces start from 1
-			multusIndices[network.Name] = len(multusIndices) + 1
+func mapNetworksByName(nets []v1.Network) map[string]v1.Network {
+	networks := map[string]v1.Network{}
+	for _, net := range nets {
+		networks[net.Name] = net
+	}
+	return networks
+}
+
+func isSecondaryMultusNetwork(net v1.Network) bool {
+	return net.Multus != nil && !net.Multus.Default
+}
+
+func getSecondaryPodInterfaceName(index int) string {
+	return fmt.Sprintf("net%d", index+1)
+}
+
+func mapPodInterfaceNameByNetwork(primaryMultusNetwork *v1.Network, secondaryMultusNetworks []v1.Network) map[string]string {
+	m := map[string]string{}
+	for i, net := range secondaryMultusNetworks {
+		m[net.Name] = getSecondaryPodInterfaceName(i)
+	}
+	if primaryMultusNetwork != nil {
+		m[primaryMultusNetwork.Name] = primaryPodInterfaceName
+	}
+	return m
+}
+
+func lookupPrimaryNetwork(nets []v1.Network) *v1.Network {
+	for _, net := range nets {
+		if !isSecondaryMultusNetwork(net) {
+			return &net
 		}
 	}
-	return networks, multusIndices
+	return nil
 }
 
-func getPodInterfaceName(networks map[string]*v1.Network, multusIndices map[string]int, ifaceName string) string {
-	if networks[ifaceName].Multus != nil && !networks[ifaceName].Multus.Default {
-		// multus pod interfaces named netX
-		return fmt.Sprintf("net%d", multusIndices[ifaceName])
-	} else {
-		return primaryPodInterfaceName
+func filterSecondaryMultusNetworks(nets []v1.Network) []v1.Network {
+	var secondary []v1.Network
+	for _, net := range nets {
+		if isSecondaryMultusNetwork(net) {
+			secondary = append(secondary, net)
+		}
 	}
+	return secondary
 }
 
 func SetupPodNetworkPhase1(vmi *v1.VirtualMachineInstance, pid int) error {
 	err := CreateVirtHandlerCacheDir(vmi.ObjectMeta.UID)
 	if err != nil {
 		return err
 	}
-	networks, multusIndices := getNetworksAndMultusIndices(vmi)
+	networks := mapNetworksByName(vmi.Spec.Networks)
+	primaryNet := lookupPrimaryNetwork(vmi.Spec.Networks)
+	secondaryNets := filterSecondaryMultusNetworks(vmi.Spec.Networks)
+	podInterfaceNames := mapPodInterfaceNameByNetwork(primaryNet, secondaryNets)
 	for i, iface := range vmi.Spec.Domain.Devices.Interfaces {
 		network, ok := networks[iface.Name]
 		if !ok {
 			return fmt.Errorf("failed to find a network %s", iface.Name)
 		}
 		podnic := podNICFactory()
-		podInterfaceName := getPodInterfaceName(networks, multusIndices, iface.Name)
-		err = podNIC.PlugPhase1(podnic, vmi, &vmi.Spec.Domain.Devices.Interfaces[i], network, podInterfaceName, pid)
+		podInterfaceName := podInterfaceNames[network.Name]
+		err = podNIC.PlugPhase1(podnic, vmi, &vmi.Spec.Domain.Devices.Interfaces[i], &network, podInterfaceName, pid)
 		if err != nil {
 			return err
 		}
```

Comment:
Can you make this identical to the one in phase2?

```
if err := podNIC.PlugPhase1(podnic, vmi, &vmi.Spec.Domain.Devices.Interfaces[i], &network, podInterfaceName, pid); err != nil {
    return err
}
```
Then the next commit/PR could extract the common code to:

```
SetupPodNetwork(...., podNIC.PlugPhase1)
SetupPodNetwork(...., podNIC.PlugPhase2)
```

Reply:
I don't understand your request. This line is different from phase2 due to the extra `pid` arg, and the fact that here the `err` is already defined above.

Reply:
Ohh I missed the pid part.
I hoped that if you use the embedded `if err := Foo(); err != {...}` on both, they will be identical.

I guess it matters much less not.